IO/d4est_norms.c

```diff
@@ -0,0 +1,617 @@
+    ctx->ghost,
+    ctx->ghost_data,
+    ctx->d4est_ops,
+    ctx->d4est_geom,
+    ctx->d4est_quad,
+    ctx->d4est_factors
+  );
+
+  // Reduce over all parallel processes
+  double norm_sq;
+  sc_reduce(
+    &norm_sq_local,
+    &norm_sq,
+    1,
+    sc_MPI_DOUBLE,
+    sc_MPI_SUM,
+    0,
+    sc_MPI_COMM_WORLD
+  );
+
+  // Need num_nodes for fit
+  int num_nodes;
+  if (ctx->fit != NULL) {
+    sc_reduce(
+      &num_nodes_local,
+      &num_nodes,
+      1,
+      sc_MPI_INT,
+      sc_MPI_SUM,
+      0,
+      sc_MPI_COMM_WORLD
+    );
+  }
+  
+  // Perform energy norm fit
+  if (ctx->fit != NULL && ctx->p4est->mpirank == 0) {
+    d4est_norms_energy_norm_add_entry_and_fit(
+      ctx->p4est,
+      ctx->fit,
+      norm_sq,
+      num_nodes
+    );
+  }
+
+  return sqrt(norm_sq);
+}
+
+double
+d4est_norms_energy_estimator
+(
+  double *field_value_errors,
+  int num_nodes_local,
+  void *norm_fcn_ctx
+)
+{
+  d4est_norms_energy_ctx_t *ctx = norm_fcn_ctx;
+  double norm_sq_local = ctx->energy_estimator_sq_local;
+
+  // Reduce over all parallel processes
+  double norm_sq;
+  sc_reduce(
+    &norm_sq_local,
+    &norm_sq,
+    1,
+    sc_MPI_DOUBLE,
+    sc_MPI_SUM,
+    0,
+    sc_MPI_COMM_WORLD
+  );
+
+  return sqrt(norm_sq);
+}
+
+
+/*
+  Computes norms over the computational grid and outputs them to files.
+  
+  - options_file: (Not implemented yet)
+  - field_names: Null terminated list of names for the fields on the computational grid.
+
+  Either provide `field_values_compare` or an `analytical_solution` for each field.
+*/
+void
+d4est_norms_save
+(
+  p4est_t *p4est,
+  const char** field_names,
+  double **field_values,
+  double **field_values_compare,
+  d4est_xyz_fcn_t *analytical_solutions,
+  void **analytical_solution_ctxs,
+  const char **norm_names,
+  d4est_norm_fcn_t *norm_fcns,
+  void **norm_fcn_ctxs
+)
+{
+  zlog_category_t *c_default = zlog_get_category("d4est_norms");
+  if (p4est->mpirank == 0)
+    zlog_debug(c_default, "Computing norms...");
+
+  // Get the number of fields and norms
+  int num_fields = -1;
+  while (field_names[++num_fields] != NULL)
+    continue;
+  int num_norms = -1;
+  while (norm_names[++num_norms] != NULL)
+    continue;
+
+  // Get number of nodes on the local process
+  int num_nodes_local = d4est_mesh_get_local_nodes(p4est);
+  int num_nodes_local_combined[2] = {
+    num_nodes_local,
+    d4est_mesh_get_local_quad_nodes(p4est)
+  };
+  int num_nodes_combined[2];
+  sc_reduce(
+    &num_nodes_local_combined,
+    &num_nodes_combined,
+    2,
+    sc_MPI_INT,
+    sc_MPI_SUM,
+    0,
+    sc_MPI_COMM_WORLD
+  );
+  int num_nodes = num_nodes_combined[0];
+  int num_quad_nodes = num_nodes_combined[1];
+  // Compute average degree
+  // TODO: can be computed in post-processing, remove?
+  int avg_deg = pow(num_nodes / p4est->global_num_quadrants, 1. / (P4EST_DIM)) - 1.f + .5f;
+  int avg_deg_quad = pow(num_quad_nodes / p4est->global_num_quadrants, 1. / (P4EST_DIM)) - 1.f + .5f;
```

Comment:
Do you want to keep this? It's so easy to compute from the data in post-processing..

Reply:
Yeah we don't need these and they clutter it up, so I'm in favour of removing.